### PR TITLE
New type reflection

### PR DESCRIPTION
**Pull request by burner**

### CODE REVIEW

File: source/graphql/helper.d [hunk 1]
```diff
@@ -858,23 +869,28 @@ Json toGraphqlJson(Schema,T)(auto ref T input) {
...
-					&& !is(types[idx] : NullableStore!Type, Type))
-			{
-				static if(is(types[idx] == enum)) {
-					ret[names[idx]] =
-						to!string(__traits(getMember, input, names[idx]));
-				} else {
-					ret[names[idx]] = toGraphqlJson!Schema(
-							__traits(getMember, input, names[idx])
-						);
+			static if(!canFind(dlangObjectMemberNames, names[idx])) {
```
[the remark this answers is not in the document]
Comment (burner) [reply source/graphql/helper.d]: thanks done

**Comment by burner** [comment]
@schveiguy can you take a look at this please and maybe also test our work project

**Comment by schveiguy** [comment]
Will do.

**Comment by schveiguy** [comment]
So I'm reviewing by reading the resulting code, since I believe lots has changed since I last looked at it, and I don't know what's important in the diff.

I'm focused on the mechanism that changes compile-time constructs into runtime type info. Looks nice and solid. Minimizing the template and CTFE calls is what we are looking for.

From what I can tell the most significant template in here is the `typeToGQLDType` function. And looks just like I would expect. The only question I have is why do the fields and then do the non-fields? I don't think combining them will affect the compile stats much, but it just looks... unnecessary.

Basically, we have 2 loops here:

```d
static foreach(idx; 0 .. fieldNames.length) {{
   ...
}

...

static foreach(mem; __traits(allMembers, Type) {{
   static if(... && isCallable!MemType) {{
      ...
   }}
}}
```
I think basically, you can put an `else static if` in there to tell if it's a field, and do the field stuff there.

Now, I think an `isField` template is not trivial to write, but could go something like:

```d
static if(is(typeof(__traits(getMember, Type, mem).offsetof)))
```

I think this might trim a bit down some of the CTFE code. I don't know how much it saves.

Reviewing the whole thing... I don't know if I know how to do that, or if it's worth it at this point. This is a big PR. I tend to think if it works and passes tests, let's do it.

**Comment by burner** [comment]
> So I'm reviewing by reading the resulting code, since I believe lots has changed since I last looked at it, and I don't know what's important in the diff.
> 
> I'm focused on the mechanism that changes compile-time constructs into runtime type info. Looks nice and solid. Minimizing the template and CTFE calls is what we are looking for.
> 
> From what I can tell the most significant template in here is the `typeToGQLDType` function. And looks just like I would expect. The only question I have is why do the fields and then do the non-fields? I don't think combining them will affect the compile stats much, but it just looks... unnecessary.
> 
> Basically, we have 2 loops here:
> 
> ```d
> static foreach(idx; 0 .. fieldNames.length) {{
>    ...
> }
> 
> ...
> 
> static foreach(mem; __traits(allMembers, Type) {{
>    static if(... && isCallable!MemType) {{
>       ...
>    }}
> }}
> ```
> 
> I think basically, you can put an `else static if` in there to tell if it's a field, and do the field stuff there.
> 
> Now, I think an `isField` template is not trivial to write, but could go something like:
> 
> ```d
> static if(is(typeof(__traits(getMember, Type, mem).offsetof)))
> ```
> 
> I think this might trim a bit down some of the CTFE code. I don't know how much it saves.
> 
> Reviewing the whole thing... I don't know if I know how to do that, or if it's worth it at this point. This is a big PR. I tend to think if it works and passes tests, let's do it.

I don't quite follow, maybe you could turn that comment into a PR.

**Pull request merged**